[Graph] Node splitting infrastructure

mciprian13 opened the pull request:
**Summary**
Added more flexible node splitting infrastructure to:
- split nodes by number of chunks
- split nodes by chunk sizes
- split nodes by chunk weights (to split a node in non-equal chunks proportional to some weights)
- any externally provided constraints (e.g. split by memory size)

Initial logic for splitting the following types of nodes:
- Conv2D, MaxPool, AvgPool, FullyConnected, MatMul, BatchMatMul, BatchedAdd, Transpose
- 10 x unary ops
- 10 x binary ops

Generalized logic for specializing InsertTensor/ExtractTensor intructions to TensorViews.

**Documentation**
None

**Test Plan**
Unit tests for options and all supported nodes.

mciprian13 commented:
@jfix71 I created a demo in this PR with the code which I think is more flexible than the one currently implemented in `parallelizeOps`. The feature I was mostly interested in was the splitting of nodes based on memory constraint (split nodes in sufficiently many pieces such that each piece does not exceed a give memory size). But the currently code can also split nodes in given number of chunks.

What I want you to do is take a look at the design and if this seems good for you I will go further with adding the logic for the rest of the operators (Conv, Pool, FC and a couple more) and add unit tests. The current logic is functional for Conv2D (I manually tested a Conv2D model with a lot of splitting configurations and validated that numerically the result is the same as the non-splitted version).

Thanks!

mciprian13 commented:
@jfix71 This is ready for you to review.
The only remaining problems are some unit tests involving the Slice operator for the OpenCL backend which do not work after generalizing the TensorView optimization in `IROptimizer.cpp` (same unit tests work for the other backends so there is something wrong with the OpenCL implementation). I would appreciate if you could help me with this since I am not familiar (at all) with the OpenCL backend.
I am fully at your disposal to have this reviewed and maybe add logic for other operators to suit your needs.
Thanks!

mciprian13 commented:
@jfix71 While you review this PR maybe you can tell me whether:
- you consider that some parts of the code can be put in some common place for other parts of Glow to use: I was thinking at the `SliceRange` utility class from `NodeSplitting.cpp` which most likely will be useful for designing a good `EliminateInsertSlice` optimization pass. Also the utility function `isSliceMemoryContiguous()` from `IROptimizer.cpp` might be useful somewhere else.
- if you consider the dimension ranges using included bounds [start, stop] (stop included) to be (or not) a lack of consistency with other parts of Glow

I am also in the process of adding a function for recursive node splitting (split a node and then all the input nodes up to a maximum depth) which will create deeper parallel computation paths in the graph and provide even more parallelization opportunities.

jfix71 commented on tools/ClassGen/NodeGen.cpp:
```diff
@@ -925,6 +925,15 @@ int main(int argc, char **argv) {
                     "Splat always keep floating point value internally but can"
                     "quantize it based on the output type.");
 
+  BB.newNode("Touch").addResultFromCtorArg().setDocstring(
+      "Generate a tensor of a specific type without initializing "
+      "it. This is useful when filling a big tensor entirely with "
+      "multiple small slices using InsertTensor nodes such that "
+      "the big tensor is not required to be initialized (filled) "
+      "with some value prior to insertion. This node is intended "
+      "to remove the overhead associated with the initialization "
+      "in situations where it is not required.");
```
(if we keep this)

mciprian13 replied:
Done.

mciprian13 commented on include/glow/Base/Type.h:
```diff
@@ -60,6 +60,12 @@ static_assert(sizeof(float16_t) == 2, "Half precision should be 16-bit");
+  static const size_t dimN = 0;
+  static const size_t dimH = 1;
+  static const size_t dimW = 2;
+  static const size_t dimC = 3;
```
Done.

mciprian13 commented on include/glow/Base/Type.h:
```diff
@@ -134,6 +140,12 @@ struct ShapeNHWTC {
+  static const size_t dimN = 0;
+  static const size_t dimC = 1;
+  static const size_t dimH = 2;
+  static const size_t dimW = 3;
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,82 @@
+
+  /// Original node which was used for splitting. When this context is provided
+  /// the original node is still part of the graph so the user can also check
+  /// graph connexion properties (like whether the original node is followed or
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,82 @@
+
+/// Function to split all the nodes from the function \p F for which there is
+/// a split logic implemented using the given split node \p constraints. When
+/// at least one of the constraints is not met the splitting is not done.
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1258 @@
+  }
+
+  // Explicitly destroy the temporary clones.
+  for (auto splitNode : splitNodes) {
```
Done.

mciprian13 commented on include/glow/Base/Type.h:
```diff
@@ -249,6 +261,10 @@ struct PaddingNFTBLR {
 struct ShapeHW {
+
+  static const size_t dimH = 0;
+  static const size_t dimW = 1;
```
Done.

mciprian13 commented on include/glow/Base/Type.h:
```diff
@@ -261,6 +277,26 @@ struct ShapeHW {
+
+  static const size_t dimN = 0;
+  static const size_t dimH = 1;
+  static const size_t dimW = 2;
```
Done.

mciprian13 commented on include/glow/Base/Type.h:
```diff
@@ -261,6 +277,26 @@ struct ShapeHW {
+    assert(shape.size() == 3 && "Invalid shape");
+    n = shape[0];
+    h = shape[1];
+    w = shape[2];
```
Done.

mciprian13 commented on include/glow/Graph/Node.h:
```diff
@@ -116,6 +111,10 @@ class Node : public Named,
   /// Dump a textual representation of the Node to std::string.
   std::string toString() const;
 
+  /// Get the total memory size (in bytes) of the node as the sum of sizes
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+  /// Ctor.
+  SplitNodeOption(std::vector<size_t> splitDims) : splitDims_(splitDims) {}
+
+  /// Getter for split dims.
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+  /// Getter for split dims.
+  std::vector<size_t> getSplitDims() const { return splitDims_; }
+
+  /// Get index of \p splitDim in the \ref splitDims_ array.
```
Done.

mciprian13 commented on lib/Graph/Node.cpp:
```diff
@@ -289,6 +291,17 @@ void Node::dump() const { dump(llvm::outs()); }
 
+size_t Node::getTotMemSize() const {
+  size_t totMemSize = 0;
+  for (unsigned idx = 0; idx < getNumInputs(); idx++) {
```
Done.

mciprian13 commented on lib/Graph/Node.cpp:
```diff
@@ -289,6 +291,17 @@ void Node::dump() const { dump(llvm::outs()); }
+  for (unsigned idx = 0; idx < getNumInputs(); idx++) {
+    totMemSize += getNthInput(idx).getType()->getSizeInBytes();
+  }
+  for (unsigned idx = 0; idx < getNumResults(); idx++) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+  /// Verify whether this range is included by another range.
+  bool isIncludedBy(const SliceRange &other) const {
+    auto rangesOther = other.getRanges();
+    assert(ranges_.size() == rangesOther.size() &&
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+  }
+
+  // Explicitly destroy the temporary nodes.
+  for (auto splitNode : splitNodes) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+        node, splitOption, splitConstraint, ConvolutionNode::ResultIdx,
+        getConv2DInputIdxAndMaps<ShapeNHWC>(dyn_cast<ConvolutionNode>(node)),
+        {}, Conv2DSplitNodeModifier<ShapeNHWC>);
+    break;
```
I removed them.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+}
+
+/// Test for SplitNodeByNumChunks option.
+TEST_F(NodeSplitting, SplitNodeByNumChunksOptionTest) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+    splitNodes[sliceIdx] = clone;
+
+    // Merge the partial outputs of this clone.
+    for (size_t outIdx = 0; outIdx < node->getNumResults(); outIdx++) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+
+  // Create split nodes.
+  std::vector<Node *> splitNodes(splitOutputSlices.size(), nullptr);
+  for (size_t sliceIdx = 0; sliceIdx < splitOutputSlices.size(); sliceIdx++) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+
+  // Allocate output tensors used for merging the partial output slices.
+  std::vector<NodeValue> mergedOutputs(node->getNumResults());
+  for (size_t outIdx = 0; outIdx < node->getNumResults(); outIdx++) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+  }
+
+  // Replace all the node outputs with the merged outputs.
+  for (size_t outIdx = 0; outIdx < node->getNumResults(); outIdx++) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+  // If default split dims are empty then return.
+  if (!splitDims.size()) {
+    return std::vector<Node *>();
+  }
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                           PlaceholderBindings &bindings,
+                           CompilationContext &cctx,
+                           std::vector<size_t> splitDims,
+                           std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                            PlaceholderBindings &bindings,
+                            CompilationContext &cctx,
+                            std::vector<size_t> splitDims,
+                            std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                             PlaceholderBindings &bindings,
+                             CompilationContext &cctx,
+                             std::vector<size_t> splitDims,
+                             std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+static void splitMatMul(Function *F, Function *&optF,
+                        PlaceholderBindings &bindings, CompilationContext &cctx,
+                        std::vector<size_t> splitDims,
+                        std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                                PlaceholderBindings &bindings,
+                                CompilationContext &cctx,
+                                std::vector<size_t> splitDims,
+                                std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                              PlaceholderBindings &bindings,
+                              CompilationContext &cctx,
+                              std::vector<size_t> splitDims,
+                              std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                                   PlaceholderBindings &bindings,
+                                   CompilationContext &cctx,
+                                   std::vector<size_t> splitDims,
+                                   std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                              PlaceholderBindings &bindings,
+                              CompilationContext &cctx,
+                              std::vector<size_t> splitDims,
+                              std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                                   PlaceholderBindings &bindings,
+                                   CompilationContext &cctx,
+                                   std::vector<size_t> splitDims,
+                                   std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                           std::vector<dim_t> outputDims,
+                           std::vector<unsigned_t> kernels,
+                           std::vector<unsigned_t> strides,
+                           std::vector<unsigned_t> pads) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                           std::vector<dim_t> outputDims,
+                           std::vector<unsigned_t> kernels,
+                           std::vector<unsigned_t> strides,
+                           std::vector<unsigned_t> pads) {
```
Done.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                             PlaceholderBindings &bindings,
+                             CompilationContext &cctx,
+                             std::vector<size_t> splitDims,
+                             std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+  bool splitNodesCheck = true;
+  std::vector<Node *> splitNodes;
+  std::list<std::unique_ptr<SliceNode>> inputSliceNodes;
+  std::list<Type> inputTypes;
+  std::list<Type> outputTypes;
```
It is about the lifetime of the objects. They must be alive at the time the split context is passed to the constraint for verification.
The reason I have this entire function using orphan nodes/types is because for some usecases (search by memory constraint) I could go through this function thousands of times and most of the times I would need to discard the inserted nodes/types with 0 reutilization.
Therefore everything that happens inside the function `verifySplitNodes` I want to happen in `orphan mode` without needing to spend extra time later for the cleaning up the Function/Module.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+    for (size_t outIdx = 0; outIdx < node->getNumResults(); outIdx++) {
+      auto nodeName =
+          clone->getName().str() + ".MergeOutput" + std::to_string(outIdx);
+      mergedOutputs[outIdx] = F->createInsertTensor(
+          nodeName, mergedOutputs[outIdx], clone->getNthResult(outIdx),
+          outputRanges[outIdx].getStarts());
```
Not quite.
The ConcatNode has the limitation of concatenating tensors along a single axis. The slices I`m dealing with here are generally split along multiple axes/dimensions. This functionality is achieved with InsertTensor nodes which are generic enough. In principle this could get achieved also with Concat nodes but would have a slightly more complex topology (multiple layers of Concats for each axis which get finally concatenated for the final result).
Another thing I don't like about Concat is that it forces/imposes the order in which the slices are concatenated/merged.
Having said that, I plan to leave this logic with InsertTensor nodes which are more flexible. A follow-up of this PR will be for me to add a good `EliminateInsertSlice` optimization pass.

jfix71 replied:
Ok, I guess it's more reasonable since you may be splitting on multiple axes. I still think it's doable to have a Concat per axis, and you'd be able to take advantage of concat optimizations. But it's not a huge deal.

mciprian13 replied:
I would leave it as is and have a follow-up PR with the `EliminateInsertSlice` optimization pass.

mciprian13 commented on tests/unittests/NodeSplittingTest.cpp:
```diff
@@ -0,0 +1,1442 @@
+                                  PlaceholderBindings &bindings,
+                                  CompilationContext &cctx,
+                                  std::vector<size_t> splitDims,
+                                  std::vector<dim_t> numChunks) {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+
+  /// Shuffle the dimension ranges using the indices \p shuffle. The flag
+  /// \p invert allows optionally to invert the shuffle permutation.
+  SliceRange shuffleRanges(std::vector<size_t> shuffle,
```
Changed to llvm::ArrayRef.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+  }
+
+  /// Getter for ranges.
+  std::vector<DimRange> getRanges() const { return ranges_; }
```
The function needs to be public. I looked over the functions again and it is ok.
Changed to llvm::ArrayRef.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+
+  /// Subscript operator for accessing a range for a given dimension.
+  DimRange &operator[](size_t dim) {
+    assert(dim < ranges_.size() && "Invalid dimension!");
```
I replaced all the asserts in the code with specialized DCHECK's.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+      << "SplitNodeByChunkSizes: Invalid number of sizes '" << numChunks
+      << "' for splitting a dimension with size '" << dimSize << "'!";
+  for (const auto &chunkSize : chunkSizes) {
+    CHECK(chunkSize > 0)
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+public:
+  /// Ctor.
+  SplitNodeByNumChunks(std::vector<size_t> splitDims,
+                       std::vector<dim_t> numChunks, bool bigChunksFirst = true)
```
Changed to llvm::ArrayRef.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+
+public:
+  /// Ctor.
+  SplitNodeOption(std::vector<size_t> splitDims) : splitDims_(splitDims) {}
```
Changed to llvm::ArrayRef.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+public:
+  /// Ctor.
+  SplitNodeByChunkSize(std::vector<size_t> splitDims,
+                       std::vector<dim_t> chunkSizes,
```
Changed to llvm::ArrayRef.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+public:
+  /// Ctor.
+  SplitNodeByChunkSizes(std::vector<size_t> splitDims,
+                        std::vector<std::vector<dim_t>> chunkSizes)
```
Changed to llvm::ArrayRef.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+public:
+  /// Ctor.
+  SplitNodeByChunkWeights(std::vector<size_t> splitDims,
+                          std::vector<std::vector<float>> chunkWeights)
```
Changed to llvm::ArrayRef.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+                       std::vector<dim_t> numChunks, bool bigChunksFirst = true)
+      : SplitNodeOption(splitDims), numChunks_(numChunks),
+        bigChunksFirst_(bigChunksFirst) {
+    CHECK(splitDims.size() == numChunks.size())
```
Done. Specialized the CHECK flavors for all the instances in the file.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+///===---------------------------------------------------------------------===//
+size_t SplitNodeOption::getSplitDimIdx(dim_t splitDim) const {
+  auto splitDimsIt = std::find(splitDims_.begin(), splitDims_.end(), splitDim);
+  CHECK(splitDimsIt != splitDims_.end())
```
Done here and elsewhere.

mciprian13 replied:
Actually when I use here `CHECK_NE` I get a compile error. I suppose the macro does not work with iterators?

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,285 @@
+/// Split node map returned by the node splitting procedure which consists in a
+/// mapping between the original node and the vector of nodes it was split into.
+/// If the node was not actually split then the vector is empty.
+using SplitNodeMap = std::unordered_map<Node *, std::vector<Node *>>;
```
I changed the pointers for option and constraint to const to forbid the node splitting functions to change them.
I left the returned map node pointer type to non-const since the user should be allowed to do anything with returned map.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1560 @@
+  /// Ctor.
+  SliceRange(TypeRef type) {
+    for (auto size : type->dims()) {
+      ranges_.emplace_back(0, size - 1);
```
I think we should decide now whether I should change the logic from [start, stop] to [start, stop). This is because my memory is fresh and it should be faster to change it now than later. Regarding the subject I don't think there is a best way to do it. I agree some functions look odd when using [start, stop] but on the other hand the logic for creating the slice range maps for Conv2D for example looks much more clean when using [start, stop] than [start, stop).
So I guess the main driving factors should be:
- consistency with other parts of Glow
- choose the most used convention
- etc.

I'll let you decide if you want me to change it. Again if you want me to change it I should do it now and not later since there are some tricky logic involved in the Conv2D which must be changed and it would be easier now.

mciprian13 replied:
Done. I changed the SliceRange representation from [start, stop] to [start, stop). I also added a new unit test to make sure I did not screw up something while changing this.

mciprian13 commented:
@jfix71 I adressed most of the review items. The main item for you to address is whether I should change the logic used in the node-splitting based on slice ranges from [start, stop] to [start, stop).

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    DCHECK_EQ(node->getNumResults(), 1) << "Binary operator invalid!";
+    return splitAndReplaceNode(
+        node, splitOption, splitConstraint, /*splitOutputIdx*/ 0,
+        {{0, CheckedSliceRangeMapIdentity}, {1, CheckedSliceRangeMapIdentity}});
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    DCHECK_EQ(node->getNumInputs(), 2) << "Binary operator invalid!";
+    DCHECK_EQ(node->getNumResults(), 1) << "Binary operator invalid!";
+    return splitAndReplaceNode(
+        node, splitOption, splitConstraint, /*splitOutputIdx*/ 0,
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,287 @@
+      : SplitNodeOption(splitDims), numChunks_(numChunks),
+        bigChunksFirst_(bigChunksFirst) {
+    CHECK_EQ(splitDims.size(), numChunks.size())
+        << "Mistmatch between 'splitDims' and 'numChunks' array sizes!";
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,287 @@
+      : SplitNodeOption(splitDims), chunkSizes_(chunkSizes),
+        bigChunksFirst_(bigChunksFirst) {
+    CHECK_EQ(splitDims.size(), chunkSizes.size())
+        << "Mistmatch between 'splitDims' and 'chunkSizes' array sizes!";
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,287 @@
+                        llvm::ArrayRef<std::vector<dim_t>> chunkSizes)
+      : SplitNodeOption(splitDims), chunkSizes_(chunkSizes) {
+    CHECK_EQ(splitDims.size(), chunkSizes.size())
+        << "Mistmatch between 'splitDims' and 'chunkSizes' array sizes!";
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,287 @@
+                          llvm::ArrayRef<std::vector<float>> chunkWeights)
+      : SplitNodeOption(splitDims), chunkWeights_(chunkWeights) {
+    CHECK_EQ(splitDims.size(), chunkWeights.size())
+        << "Mistmatch between 'splitDims' and 'chunkWeights' array sizes!";
```
Done.

mciprian13 commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,287 @@
+  /// nodes attached to their outputs so you are only allowed to check the split
+  /// nodes types, attributes, input/output operand types and also the offsets
+  /// of the SliceNodes used for splitting the input operands.
+  std::vector<Node *> splitNodes;
```
I used a struct since I thought maybe we would want to add more information in the future. But I followed your advice and I removed the struct entirely and used instead a `const Node *origNode` and a `const std::vector<Node *> &splitNodes` directly in the constraint API.

mciprian13 commented on lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:
```diff
@@ -4567,6 +4577,14 @@ Error glow::optimizeFunction(Function *F, const Backend &B,
+        splitMap, ::glow::splitNodes(
+                      F, SplitNodeMaxMemConstraint(splitNodesByMemSizeOpt)));
+  }
+
```
You're right. I added it there for some particular tests and forgot it.
I removed it.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    }
+  }
+
+  /// Getter for ranges.
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    return sizes;
+  }
+
+  /// Get number of dimensions.
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    return starts;
+  }
+
+  /// Getter for dimensions sizes.
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+  /// Get number of dimensions.
+  size_t getNumDims() const { return ranges_.size(); }
+
+  /// Subscript operator for accessing a range for a given dimension.
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    return ranges_[dim];
+  }
+
+  const DimRange &operator[](size_t dim) const {
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    return SliceRange(dimRanges);
+  }
+
+  /// Verify that slice range is valid (non-empty).
```
Done. I also inverted the logic by renaming the function `isEmpty` which should have a more obvious meaning than `isValid`.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+  }
+
+  /// Shuffle the dimension ranges using the indices \p shuffle. The flag
+  /// \p invert allows optionally to invert the shuffle permutation.
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    return true;
+  }
+
+  /// Verify that both ends of a dimension \p dim are aligned to \p align. For
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+           ((ranges_[dim].second + 1) % align == 0);
+  }
+
+  /// Verify whether this range is included by another range.
```
Done.

mciprian13 commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1556 @@
+    return true;
+  }
+
+  /// Get a textual representation.
```
Done.

mciprian13 commented:
@jfix71 I created a separate PR #4497 for the IROptimization.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+class SplitNodeOption {
+
+  /// All the dimensions of a tensor used for splitting in increasing order.
+  std::vector<size_t> splitDims_;
```
Could you use a `llvm::SmallVector<size_t, max_tensor_dimensions>` to avoid dynamic memory allocation?

mciprian13 replied:
Question: The `max_tensor_dimensions` does limit the maximum number of elements or is just used for initially allocating the vector?

opti-mix replied:
It initially reserves memory for `most max_tensor_dimensions` on stack. If at runtime more elements are needed, it will grow and use heap allocation, just like a usual std::vector. You usually use SmallVector if you know you are going to have a small number of elements in your vector and you know the likely high watermark for this number in advance.

mciprian13 replied:
Got it. I started making the replacements and I got a lot of errors. How do I initialize a SmallVector with an ArrayRef here:
```
SplitNodeOption(llvm::ArrayRef<size_t> splitDims) : splitDims_(splitDims) {}
```

opti-mix replied:
I guess, you should be able to use iterators: `splitDims_(splitDims.begin(), splitDims.end())`

mciprian13 replied:
Done.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  virtual std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const = 0;
```
Should the return type be a mutable vector? If not, return an ArrayRef.

mciprian13 replied:
Stupid question: if inside this function I use a local `std::vector` and return it as a `llvm::ArrayRef` which you said is nothing but a reference, wouldn't the `ArrayRef` remain dangling after the local `std::vector` is destroyed? You have to bare with me cuz I'm not an experienced C++ programmer ...

opti-mix replied:
If you create a new vector inside the function you cannot use ArrayRef to return it. You really need to return std::vector.

Alternatively, you pass an output `llvm::SmallVectorImpl<size_t> &out` parameter to the function and add your elements to this vector. In this case the function does not need to return anything as it is done through this out parameter.

mciprian13 replied:
Ok it makes sense now. So for these functions I will preserve `std::vector` since these generate new vectors. Thanks!

mciprian13 replied:
Mutable vector so I left it `std::vector`.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+class SplitNodeByNumChunks : public SplitNodeOption {
+
+  /// Number of chunks for each split dimension.
+  std::vector<dim_t> numChunks_;
```
SmallVector?

mciprian13 replied:
Done.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
```
Should the return type be a mutable vector? If not, return an ArrayRef.

mciprian13 replied:
Mutable vector so I left it `std::vector`.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+class SplitNodeByChunkSize : public SplitNodeOption {
+
+  /// Chunk size for each split dimension.
+  std::vector<dim_t> chunkSizes_;
```
SmallVector?

mciprian13 replied:
Done.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
```
Should the return type be a mutable vector? If not, return an ArrayRef.

mciprian13 replied:
Mutable vector so I left it `std::vector`.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+class SplitNodeByChunkSizes : public SplitNodeOption {
+
+  /// Array of chunk sizes for each split dimension.
+  std::vector<std::vector<dim_t>> chunkSizes_;
```
SmallVector?

mciprian13 replied:
Done.

opti-mix commented on include/glow/Optimizer/GraphOptimizer/NodeSplitting.h:
```diff
@@ -0,0 +1,283 @@
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
```
Should the return type be a mutable vector? If not, return an ArrayRef.

mciprian13 replied:
Mutable vector so I left it `std::vector`.

opti-mix commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1561 @@
+class SliceRange {
+
+  /// Vector of ranges for all the dimensions of a slice.
+  std::vector<DimRange> ranges_;
```
SmallVector?

mciprian13 replied:
Done.

opti-mix commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1561 @@
+/// Utility function to split an array of slice ranges \p ranges along the given
+/// dimension \p dim using the split option \p splitOption.
+static std::vector<SliceRange>
+splitSliceRanges(const std::vector<SliceRange> &ranges, size_t dim,
```
ArrayRef?

mciprian13 replied:
Mutable vector so I left it `std::vector`.

opti-mix commented on tools/ClassGen/InstrGen.cpp:
```diff
@@ -671,7 +671,8 @@ int main(int argc, char **argv) {
       .addOperand("Dest", OperandKind::Out)
       .dataParallel()
-      .autoVerify(VerifyKind::NoVerify);
+      .autoVerify(VerifyKind::NoVerify)
```
Should InstrGen and NodeGen changes be moved into a separate PR? These changes seem to be pretty independent of the rest of this PR.

mciprian13 replied:
I use the TouchNode in this PR so I'm not sure why I should put this in another PR ...

opti-mix commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1561 @@
+    const SplitNodeModifier &splitNodeModifier = SplitNodeModifierNop) {
+
+  // If the split output operand has no dimensions then return.
+  if (!node->getType(splitOutputIdx)->dims().size()) {
```
May be use `empty` instead of `!size()`?

mciprian13 replied:
Done.

opti-mix commented:
@mciprian13 Could you provide a high-level overview of the splitting? 

E.g. which invariants are preserved after splitting? What is guaranteed?

Most of the tests are at the graph level. There are no executable tests that run a model if I see it correctly. Is it guaranteed that model execution results will be the same after splitting? If this is the case, should there be some tests to prove it? 

Are there any operations that cannot be split as it would result in incorrect computation results? 

What about splits with overlaps? Are they (planned to be) supported? E.g. if you split a sequence of convolutions into say 4 tiles and then you try to fuse/pipe same tiles of each of those convs (like conv1tile1, conv2tile1, conv3tile1,...) to run on the same core/processing unit, you'd need to either support halo rows exchange or have overlapping splits, so that no halo rows exchange is required. Do you envision supporting something like that in the future?

mciprian13 commented:
@opti-mix The unit tests for `NodeSplitting` use same infrastructure as `GraphOptz` to check numerical equivalence through `checkNumericalEquivalence()`. I actually use  `checkNumericalEquivalence(0)` since I expect bit-exact results.

Currently there is support for splitting 28 operators. Mathematically not all the operators can be split (e.g. `Softmax`) so node splitting must be interpreted as a best effort approach. Why I did here is something which can be scaled easily for other operators as well. The split logic must be explicitly expressed for each operator individually, but it is minimal.

The infrastructure is pretty capable at this point: it correctly handles splitting tensors non-orthogonally, for example if I want to split the Conv2D output in 2 pieces along H/W dimensions, this will result in 2 input pieces which overlap. I use some functional mapping between the input and output operands of the nodes so the infrastructure knows if I want to compute a subtensor of the output what input subtensor it corresponds to.

After this PR I will also add support for recursive splitting: split a node with some options, then apply splitting to its input operands with the same split boundaries such that finally it will result multiple completely parallel computation paths (subgraphs) in the graph.

mciprian13 commented:
@opti-mix Thanks for the review.
@jfix71 I think I addressed all the review items you raised. Have a second look and let me know.

As a follow-up I plan to do after this PR another one including:
- an `EliminateInsertSlice` optimization pass
- a recursive node splitting function to enhance the current infrastructure
- these two items go hand in hand (coupled unit tests) so I want to do them together

mciprian13 commented:
As a bonus:
- After you merge the definitions of Conv2D and Conv3D I will also add logic for splitting Conv3D which should be fairly simple.
- After #4350 gets landed (hopefully soon) I will also add logic for splitting `ChannelwiseQuantizedConv`

opti-mix commented on include/glow/Base/Type.h:
```diff
@@ -60,17 +60,23 @@ static_assert(sizeof(float16_t) == 2, "Half precision should be 16-bit");
 
 struct ShapeNHWC {
+
+  static constexpr size_t dimN = 0;
```
Minor nit: Why not just define an enum for these `dimX` constants?

mciprian13 replied:
Like this?
```
struct ShapeNHWC {

  enum {
    dimN = 0,
    dimH = 1,
    dimW = 2,
    dimC = 3,
  };
```

opti-mix replied:
Yes, exactly! I think you don't even need to explicitly assign the values to them as enum declaration does exactly this anyways. Also, as a very tiny nit, enum cases are usually starting with an upper-case latter.

mciprian13 replied:
Oh boy ... I'll need to modify in a lot of places. Just to confirm before I embark in this journey: you mean `DimN` right?

opti-mix replied:
Yes, I meant `DimN` ;-) But I don't insist on this renaming. It is really a tiny nit. If it is too much work, ignore this renaming part.

mciprian13 replied:
Done.

opti-mix commented on lib/Optimizer/GraphOptimizer/NodeSplitting.cpp:
```diff
@@ -0,0 +1,1561 @@
+///===---------------------------------------------------------------------===//
+///                                    Conv
+///===---------------------------------------------------------------------===//
+static std::tuple<bool, DimRange, DimPads>
```
I'd suggest defining a struct with these fields and semantically meaningful fields names and return it here. It will make the code easier to understand.

mciprian13 replied:
Done.

opti-mix commented:
Yes, I think it looks pretty good now! Let's land it!

@mciprian13 This is a great feature! Looking forward to more PRs building on top of this one!

jfix71 commented:
@mciprian13 Were you intending on adding documentation in this PR before landing?

opti-mix commented:
> Were you intending on adding documentation in this PR before landing?

Yes, some documentation about this pretty big feature would surely help!

mciprian13 commented:
@opti-mix @jfix71 What I propose is to have this landed and I will add documentation with the next PR which will actually make this feature official.
